Add MXRT595 support

mmahadevan108 opened the pull request:
Add basic support for MXRT595 EVK board. This support includes UART, GPIO support. Tested using Hello World, Blinky and Button test applications.

mmahadevan108 committed:
boards: Add support for NXP board mimxt595_evk

Add support for mimxrt595_evk

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: Add support for MXRT595

Add support for MXRT595

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: Add support for NXP MXRT5xx

Add support for MXRT595S

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: Update GPIO port numbers

Add more GPIO port numbers to the enum list

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
driver: clock: Update MCUX Syscon clock control driver

1. Update to add support for Flexcomm8-13.
2. Fix the clock control driver, the enclosing #define
   was incorrect.
3. Identify HS_SPI port using the appropriate Register
   define

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
west.yml: Update NXP HAL to add support for MIMXRT595

Add support for MIMXRT595

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

hakehuang commented:
@mmahadevan108 build testing OK, board CI is running will feedback tomorrow

gmarull commented on soc/arm/nxp_imx/rt5xx/soc.h:
```diff
@@ -0,0 +1,78 @@
+#ifndef _SOC__H_
+#define _SOC__H_
+
+#ifndef _ASMLANGUAGE
+#include <sys/util.h>
+#include <fsl_common.h>
+
+/* Add include for DTS generated information */
+#include <devicetree.h>
```
none of these headers is required here.

mmahadevan108 replied:
Build breaks if I delete any of these headers.

gmarull replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/41544 and https://github.com/zephyrproject-rtos/zephyr/issues/41543

mmahadevan108 replied:
Done, fixed.

gmarull commented on soc/arm/nxp_imx/rt5xx/soc.c:
```diff
@@ -0,0 +1,115 @@
+ * hardware for the RT5XX platforms.
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/uart.h>
+#include <linker/sections.h>
+#include <arch/cpu.h>
+#include <aarch32/cortex_m/exc.h>
+#include <fsl_clock.h>
+#include <fsl_common.h>
+#include <fsl_device_registers.h>
+#include <clock_config.h>
```
please review include list here, e.g. why `drivers/uart.h` here?

mmahadevan108 replied:
Thanks for pointing this, I have trimmed this list of includes.

gmarull commented on dts/arm/nxp/nxp_rt5xx_common.dtsi:
```diff
@@ -0,0 +1,487 @@
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <2>;
+	};
+
+	gpio3: gpio@3 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x100000 0x1000>;
+		label = "GPIO_3";
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <3>;
+	};
+
+	gpio4: gpio@4 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x100000 0x1000>;
+		label = "GPIO_4";
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <4>;
+	};
+
+	gpio5: gpio@5 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x100000 0x1000>;
+		label = "GPIO_5";
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <5>;
+	};
+
+	gpio6: gpio@6 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x100000 0x1000>;
+		label = "GPIO_6";
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <6>;
+	};
```
question: why do all GPIOs share the same reg? what is the point of `gpio@N` then?

dleach02 replied:
The GPIO on this part is located at 0x<n>010_0000 where <n> is the offset for secure/ns access. The access to the various GPIO instances is via a port offset written into the register where we have values 0 through 6 for this part. The same thing is present in the rt6xx and other LPC based IP blocks.

gmarull commented on dts/arm/nxp/nxp_rt5xx_common.dtsi:
```diff
@@ -0,0 +1,487 @@
+
+	hs_lspi: spi@126000 {
+		compatible = "nxp,lpc-spi";
+		/* Enabling cs-gpios below will allow using GPIO CS,
+		 rather than Flexcomm SS */
+		/* cs-gpios = <&gpio1 14 GPIO_ACTIVE_LOW>,
+			<&gpio1 15 GPIO_ACTIVE_LOW>,
+			<&gpio1 16 GPIO_ACTIVE_LOW>,
+			<&gpio1 17 GPIO_ACTIVE_LOW>; */
```
please document this on the DT bindings, not as a comment

mmahadevan108 replied:
I am going to delete these lines of code. They are in place in case we want to switch to using GPIO instead of the Slave Select from the IP block.

gmarull commented on dts/arm/nxp/nxp_rt5xx.dtsi:
```diff
@@ -0,0 +1,25 @@
+	};
+};
+
+#include "nxp_rt5xx_common.dtsi"
```
any particular reason to include at the bottom?

mmahadevan108 replied:
Some of the nodes defined here are used by the `_common` file.

gmarull commented on include/dt-bindings/clock/mcux_lpc_syscon_clock.h:
```diff
@@ -15,12 +15,19 @@
+#define MCUX_FLEXCOMM11_CLK		11
+#define MCUX_FLEXCOMM12_CLK		12
+#define MCUX_FLEXCOMM13_CLK		13
+#define MCUX_HS_SPI_CLK		14
+#define MCUX_PMIC_I2C_CLK		15
 
-#define MCUX_CTIMER_CLK_OFFSET  11
+#define MCUX_USDHC1_CLK		20
+#define MCUX_USDHC2_CLK		21
+
+#define MCUX_CTIMER_CLK_OFFSET  22
```
nit: mixes tabs/spaces, please align properly

mmahadevan108 replied:
Thanks for pointing this, fixed.

gmarull commented on west.yml:
```diff
@@ -98,7 +98,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 66db7c2a57ec1adfd3574169346ed3af9bf17c0b
+      revision: pull/131/head
```
this commit should go first for bisectability

mmahadevan108 replied:
Thanks, done.

gmarull commented on boards/arm/mimxrt595_evk/board.h:
```diff
@@ -0,0 +1,18 @@
+#include "clock_config.h"
+#include "fsl_common.h"
+
+void BOARD_FlexspiClockSafeConfig(void);
+void BOARD_SetFlexspiClock(FLEXSPI_Type *base, uint32_t src, uint32_t divider);
+void BOARD_DeinitFlash(FLEXSPI_Type *base);
+void BOARD_InitFlash(FLEXSPI_Type *base);
```
are these functions public? why don't they follow `lower_case()` naming? I don't think we should be accepting this sort of code at board level, except this is a temporary solution until proper drivers/DT support is in place.

mmahadevan108 replied:
Thank you for catching this, I have update the function names. This is how I see it implemented in our SDK, I have moved these functions to the `soc` folder with different function names.

hakehuang commented:
@mmahadevan108 
```
tests/arch/arm/arm_irq_vector_table

    E: ***** MPU FAULT *****

    E:   Instruction Access Violation

    E: r0/a1:  0x18004b71  r1/a2:  0x00000000  r2/a3:  0x18004b71

    E: r3/a4:  0x301816a8 r12/ip:  0x00000000 r14/lr:  0xfffffffd

    E:  xpsr:  0x68000039

    E: Faulting instruction address (r15/pc): 0x42E: ***** MPU FAULT *****

    E:   Instruction Access Violation

    E: r0/a1:  0x18004b71  r1/a2:  0x00000000  r2/a3:  0x18004b71

    E: r3/a4:  0x301816a8 r12/ip:  0x00000000 r14/lr:  0xfffffffd

    E:  xpsr:  0x68000039

    E: Faulting instruction address (r15/pc): 0x42984b16

    E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0

    E: Fault during interrupt handling

    

    E: Current thread: 0x30180ef0 (main)

    E: Halting system
```

gmarull commented on soc/arm/nxp_imx/rt5xx/Kconfig.defconfig.mimxrt595_cm33:
```diff
@@ -0,0 +1,41 @@
+config SOC
+	default "mimxrt595s_cm33"
+
+config PINMUX_MCUX_LPC
+	default y
+	depends on PINMUX
```
note: all pinmux seems to happen at board level, is that really needed?

mmahadevan108 replied:
This is so that we can enable the PINMUX driver.

gmarull commented on soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:
```diff
@@ -0,0 +1,141 @@
+		}
+	} else if (base == FLEXSPI1) {
+		if ((CLKCTL0->FLEXSPI1FCLKSEL != CLKCTL0_FLEXSPI1FCLKSEL_SEL(src)) ||
+		    ((CLKCTL0->FLEXSPI1FCLKDIV & CLKCTL0_FLEXSPI1FCLKDIV_DIV_MASK) != (divider - 1))) {
+			/* Always deinit FLEXSPI and init FLEXSPI for the flash to make sure the flash works correctly after the
+			 * FLEXSPI root clock changed as the default FLEXSPI configuration may does not work for the new root clock
+			 * frequency.
+			 */
+			flash_deinit(base);
```
code in this file seems to exceed the maximum width (100cols). I'm not familiar with this, but shouldn't a clock control driver be implemented?

gmarull commented on soc/arm/nxp_imx/rt5xx/linker.ld:
```diff
@@ -0,0 +1,15 @@
+ * This is the linker script for both standard images and XIP images.
+ */
+
+#include <autoconf.h>
```
this header is not used in this file

mmahadevan108 replied:
Deleted, thank you.

gmarull commented on soc/arm/nxp_imx/rt5xx/soc.h:
```diff
@@ -0,0 +1,75 @@
+#define _SOC__H_
+
+#ifndef _ASMLANGUAGE
+#include <sys/util.h>
+#include <fsl_common.h>
```
these are not used here: files that require these headers should just include them.

mmahadevan108 replied:
This is a bigger change as there a number of files in Zephyr including this file. Maybe this can be addressed as a separate PR.

mmahadevan108 commented:
I am unable to understand the ci failures. Any idea what I am missing?
```
Python exception in `./scripts/ci/check_compliance.py`:

Traceback (most recent call last):
  File "./scripts/ci/check_compliance.py", line 1091, in main
    n_fails = _main(args)
  File "./scripts/ci/check_compliance.py", line 1042, in _main
    test.run()
  File "./scripts/ci/check_compliance.py", line 721, in run
    self.check_source_file(fname)
  File "./scripts/ci/check_compliance.py", line 777, in check_source_file
    contents = f.read()
  File "/usr/lib/python3.8/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x94 in position 1375: invalid start byte
```

mmahadevan108 commented:
The script `scripts/ci/check_compliance.py` runs successfully without errors when I run it locally.

dleach02 commented:
@mmahadevan108 need to update west.yaml

dleach02 commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -76,18 +71,36 @@ static int mcux_lpc_syscon_clock_control_get_subsys_rate(
 		break;
 	case MCUX_HS_SPI_CLK:
-#if defined(FSL_FEATURE_FLEXCOMM8_SPI_INDEX)
+#if (FSL_FEATURE_SOC_FLEXCOMM_COUNT == 9)
```
"9"? Bit of a magic number?

mmahadevan108 replied:
Thanks for spotting this. I have changed the implementation to use HSLSPI register defines.

mmahadevan108 committed:
tests: arch: arm_irq_vector_table: Update to run on MXRT595

Update the custom vector table to add the OS Event timer
interrupt which is used on RT595 as the kernel system timer

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -0,0 +1,196 @@
+
+/dts-v1/;
+
+#include <nxp/nxp_rt5xx.dtsi>
```
Done, thank you.

mmahadevan108 commented:
> @mmahadevan108 need to update west.yaml

Done.

mmahadevan108 commented:
> I am unable to understand the ci failures. Any idea what I am missing?
> 
> ```
> Python exception in `./scripts/ci/check_compliance.py`:
> 
> Traceback (most recent call last):
>   File "./scripts/ci/check_compliance.py", line 1091, in main
>     n_fails = _main(args)
>   File "./scripts/ci/check_compliance.py", line 1042, in _main
>     test.run()
>   File "./scripts/ci/check_compliance.py", line 721, in run
>     self.check_source_file(fname)
>   File "./scripts/ci/check_compliance.py", line 777, in check_source_file
>     contents = f.read()
>   File "/usr/lib/python3.8/codecs.py", line 322, in decode
>     (result, consumed) = self._buffer_decode(data, self.errors, final)
> UnicodeDecodeError: 'utf-8' codec can't decode byte 0x94 in position 1375: invalid start byte
> ```

Updated the test

dleach02 commented:
@hakehuang can you do a test run on this PR on a rt5xx board?

hakehuang commented:
@chen-png  can you check this up_squared board result?
```
Verifying content of results/v2.7.99-3293-g93b0ea978293/up_squared.xml
python3 scripts/results_verification.py -Z v2.7.99-3293-g93b0ea978293 -P ./results/v2.7.99-3293-g93b0ea978293/up_squared.xml -E 500 -F 500 -S 10
Traceback (most recent call last):
  File "scripts/results_verification.py", line 121, in <module>
    main(args)
  File "scripts/results_verification.py", line 103, in main
    assert check_name(file_path), "Report name does not match the platform name given in the report"
  File "scripts/results_verification.py", line 46, in check_name
    summary = ET.parse(file_path).getroot()[0]
  File "/usr/lib/python3.8/xml/etree/ElementTree.py", line 1202, in parse
    tree.parse(source, parser)
  File "/usr/lib/python3.8/xml/etree/ElementTree.py", line 595, in parse
    self._root = parser._parse_whole(source)
xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 590254, column 4126
Error: Process completed with exit code 1.
```

hakehuang commented:
@dleach02 CI is done, all PASS, but there is another issue with the ttests/arch/arm/arm_irq_vector_table, when I the software irq is triggerred, there is no console output, but this shall be a test issue, I will try to analysis this with Mahesh.

mmahadevan108 commented:
Below is the serial output for the `tests\arch\arm\arm_irq_vector_table`

```
Running test suite vector_table
===================================================================
START - test_arm_irq_vector_table
Test Cortex-M IRQs installed directly in the vector table
isr0 ran!
isr1 ran!
isr2 ran!
 PASS - test_arm_irq_vector_table in 0.9 seconds
===================================================================
Test suite vector_table succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
